src/openrct2/drawing/Drawing.h

```diff
@@ -230,6 +230,181 @@ struct rct_size16
+        result._tertiary = tertiary & 0xFF;
+        return result;
+    }
+
+    ImageId() = default;
+
+    explicit ImageId(uint32_t index)
+        : _value(index & MASK_INDEX)
+    {
+    }
+
+    ImageId(uint32_t index, uint8_t primaryColourOrPalette)
+        : ImageId(ImageId(index).WithPrimary(primaryColourOrPalette))
+    {
+    }
+
+    ImageId(uint32_t index, colour_t primaryColour, colour_t secondaryColour)
+        : ImageId(ImageId(index).WithPrimary(primaryColour).WithSecondary(secondaryColour))
+    {
+    }
+
+    ImageId(uint32_t index, colour_t primaryColour, colour_t secondaryColour, colour_t tertiaryColour)
+        : ImageId(ImageId(index).WithPrimary(primaryColour).WithSecondary(secondaryColour).WithTertiary(tertiaryColour))
+    {
+    }
+
+    uint32_t ToUInt32() const
+    {
+        return _value;
+    }
+
+    bool HasValue() const
+    {
+        return GetIndex() != INDEX_UNDEFINED;
+    }
+
+    bool HasPrimary() const
+    {
+        return (_value & FLAG_PRIMARY) || (_value & FLAG_SECONDARY);
+    }
+
+    bool HasSecondary() const
+    {
+        return _value & FLAG_SECONDARY;
+    }
+
+    bool HasTertiary() const
+    {
+        return !(_value & FLAG_PRIMARY) && (_value & FLAG_SECONDARY);
+    }
+
+    bool IsRemap() const
+    {
+        return (_value & FLAG_PRIMARY) && !(_value & FLAG_SECONDARY);
+    }
+
+    bool IsBlended() const
+    {
+        return _value & FLAG_BLEND;
+    }
+
+    uint32_t GetIndex() const
+    {
+        return _value & MASK_INDEX;
+    }
+
+    uint8_t GetRemap() const
+    {
+        return (_value & MASK_REMAP) >> SHIFT_REMAP;
+    }
+
+    colour_t GetPrimary() const
+    {
+        return (_value & MASK_PRIMARY) >> SHIFT_PRIMARY;
+    }
+
+    colour_t GetSecondary() const
+    {
+        return (_value & MASK_SECONDARY) >> SHIFT_SECONDARY;
+    }
+
+    colour_t GetTertiary() const
+    {
+        return _tertiary;
+    }
+
+    ImageCatalogue GetCatalogue() const;
+
+    ImageId WithIndex(uint32_t index)
+    {
+        ImageId result = *this;
+        result._value = (_value & ~MASK_INDEX) | (index & MASK_INDEX);
+        return result;
+    }
+
+    ImageId WithPrimary(colour_t colour)
+    {
+        ImageId result = *this;
+        result._value = (_value & ~MASK_PRIMARY) | ((colour << SHIFT_PRIMARY) & MASK_PRIMARY) | FLAG_PRIMARY;
+        return result;
+    }
+
+    ImageId WithSecondary(colour_t colour)
+    {
+        ImageId result = *this;
+        result._value = (_value & ~MASK_SECONDARY) | ((colour << SHIFT_SECONDARY) & MASK_SECONDARY) | FLAG_SECONDARY;
+        return result;
+    }
+
+    ImageId WithTertiary(colour_t tertiary)
+    {
+        ImageId result = *this;
+        result._value &= ~FLAG_PRIMARY;
+        if (!(_value & FLAG_SECONDARY))
+        {
+            // Tertiary implies primary and secondary, so if colour was remap (8-bit primary) then
+            // we need to zero the secondary colour.
+            result._value &= ~MASK_SECONDARY;
```

Comment:
Is this correct. Secondary mask would overlap remap mask

Reply:
I think so, you can't have a remap colour as well as a secondary or tertiary colour. It will zero some of the high bits of the remap colour.